File: Packages/TeamTasksKit/Sources/TeamTasksCore/ViewModels/Groups/CreateGroupViewModel.swift
import Foundation
import Observation

/// « Nouveau groupe » sheet (also the « Créer » side of the onboarding's first group). On success `createdGroup` is
/// set: dismiss and show it (`router.showGroup(group.id)`). The creator is the group's admin.
///
/// v2: the group's color (every `ColorKey`) and emoji (`EmojiChoices.groups`, or none: the initials), with a live
/// `preview`. The color starts on a palette color picked at random (like an automatic color) and is always sent, so
/// that the group looks as previewed.
@MainActor
@Observable
public final class CreateGroupViewModel: ErrorPresenting {
    public static let maxNameLength = Limits.groupName.upperBound
    /// Example names, for the field's placeholder.
    public static let namePlaceholder = "Coloc’, famille, asso…"

    public var name: String {
        get { nameValue }
        set {
            nameValue = newValue
            if nameError != nil { nameError = Self.nameMessage(newValue) }
        }
    }

    /// v2: the group's color.
    public var color: ColorKey
    /// v2: the group's emoji; nil = none (the initials are shown).
    public var emoji: String?

    public private(set) var nameError: String?
    public private(set) var isSubmitting = false
    public private(set) var createdGroup: GroupSummary?
    public var error: ErrorState?

    public let session: SessionModel
    private var nameValue = ""

    /// - Parameters:
    ///   - color: the initial color; nil: a palette color picked at random.
    ///   - emoji: the initial emoji; nil: none.
    public init(session: SessionModel, color: ColorKey? = nil, emoji: String? = nil) {
        self.session = session
        self.color = color ?? ColorKey.automatic(for: UUID())
        self.emoji = emoji
    }

    public var canSubmit: Bool { !InputValidation.trimmed(nameValue).isEmpty && !isSubmitting }

    // MARK: - Appearance (v2)

    public var colorOptions: [ColorKey] { ColorKey.allCases }
    public var emojiOptions: [String] { EmojiChoices.groups }

    /// The group as it will look (initials of the name typed so far).
    public var preview: AvatarAppearance {
        AvatarAppearance(color: color, emoji: emoji, initials: Initials.of(nameValue))
    }

    public func selectColor(_ option: ColorKey) {
        color = option
    }

    /// Picks an emoji; picking the selected one again removes it (nil: none).
    public func selectEmoji(_ option: String?) {
        emoji = option == emoji ? nil : option
    }

    // MARK: - Create

    /// Creates the group with its appearance. Returns it on success, nil otherwise (`nameError` or `error` set).
    @discardableResult
    public func create() async -> GroupSummary? {
        guard !isSubmitting else { return nil }
        error = nil
        nameError = Self.nameMessage(nameValue)
        guard nameError == nil else { return nil }
        isSubmitting = true
        defer { isSubmitting = false }
        do {
            let group = try await session.services.groups.createGroup(name: nameValue, color: color, emoji: emoji)
            createdGroup = group
            session.feed.bumpMemberships()
            return group
        } catch {
            guard let appError = ErrorState(from: error)?.error else { return nil }
            if appError == .invalidName {
                nameError = appError.messageFR
            } else {
                present(appError)
            }
            return nil
        }
    }

    public static func nameMessage(_ name: String) -> String? {
        do {
            _ = try InputValidation.groupName(name)
            return nil
        } catch {
            return AppError.invalidName.messageFR
        }
    }
}

File: android/core/src/main/kotlin/io/github/notkanaa/equipe/core/viewmodel/CreateGroupViewModel.kt
package io.github.notkanaa.equipe.core.viewmodel

import io.github.notkanaa.equipe.core.AppError
import io.github.notkanaa.equipe.core.GroupSummary
import io.github.notkanaa.equipe.core.InputValidation
import io.github.notkanaa.equipe.core.Limits
import kotlin.coroutines.cancellation.CancellationException

// Port of TeamTasksCore/ViewModels/Groups/CreateGroupViewModel.swift.

/** State of the « Nouveau groupe » sheet. */
data class CreateGroupState(
    val name: String = "",
    val nameError: String? = null,
    val isSubmitting: Boolean = false,
    /** Set on success: dismiss and show it (`router.showGroup(group.id)`). */
    val createdGroup: GroupSummary? = null,
    override val error: ErrorState? = null,
) : ErrorHolder {
    val canSubmit: Boolean get() = InputValidation.trimmed(name).isNotEmpty() && !isSubmitting
}

/**
 * « Nouveau groupe » sheet. On success [CreateGroupState.createdGroup] is set: dismiss and show it
 * (`router.showGroup(group.id)`). The creator is the group's admin.
 */
class CreateGroupViewModel(
    val session: SessionModel,
) : ScreenModel<CreateGroupState>(CreateGroupState(), { state, error -> state.copy(error = error) }) {
    var name: String
        get() = current.name
        set(value) = update { it.copy(name = value, nameError = if (it.nameError != null) nameMessage(value) else null) }

    /** Creates the group. Returns it on success, null otherwise (`nameError` or `error` set). */
    suspend fun create(): GroupSummary? {
        if (current.isSubmitting) return null
        dismissError()
        val nameError = nameMessage(current.name)
        update { it.copy(nameError = nameError) }
        if (nameError != null) return null
        update { it.copy(isSubmitting = true) }
        try {
            val group = session.services.groups.createGroup(current.name)
            update { it.copy(createdGroup = group) }
            session.feed.bumpMemberships()
            return group
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            val appError = ErrorState.from(error)?.error ?: return null
            if (appError == AppError.InvalidName) {
                update { it.copy(nameError = appError.messageFR) }
            } else {
                present(appError)
            }
            return null
        } finally {
            update { it.copy(isSubmitting = false) }
        }
    }

    companion object {
        /** 60 (code points). */
        val MAX_NAME_LENGTH: Int = Limits.groupName.last

        fun nameMessage(name: String): String? = try {
            InputValidation.groupName(name)
            null
        } catch (error: AppError) {
            AppError.InvalidName.messageFR
        }
    }
}
